nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/EnforceOrder.java

```diff
@@ -0,0 +1,543 @@
+            flowFileGroups.entrySet().stream().filter(entry -> !entry.getValue().isEmpty()).map(entry -> {
+                // Sort flow files within each group.
+                final List<FlowFile> groupedFlowFiles = entry.getValue();
+                groupedFlowFiles.sort(Comparator.comparing(getOrder));
+                return entry;
+            }).forEach(entry -> {
+                // Check current state.
+                final String groupId = entry.getKey();
+                final String stateKeyOrder = STATE_TARGET_ORDER.apply(groupId);
+                final int previousTargetOrder = Integer.parseInt(groupStates.get(stateKeyOrder));
+                final AtomicInteger targetOrder = new AtomicInteger(previousTargetOrder);
+                final List<FlowFile> groupedFlowFiles = entry.getValue();
+                final String maxOrderStr = groupStates.get(STATE_MAX_ORDER.apply(groupId));
+
+                groupedFlowFiles.forEach(f -> {
+                    final Integer order = getOrder.apply(f);
+                    final boolean isMaxOrder = !isBlank(maxOrderStr) && order.equals(Integer.parseInt(maxOrderStr));
+
+                    if (order == targetOrder.get()) {
+                        transferResult(f, REL_SUCCESS, null, null);
+                        if (!isMaxOrder) {
+                            // If max order is specified and this FlowFile has the max order, don't increment target anymore.
+                            targetOrder.incrementAndGet();
+                        }
+
+                    } else if (order > targetOrder.get()) {
+
+                        if (now - Long.parseLong(f.getAttribute(ATTR_STARTED_AT)) > waitTimeoutMillis) {
+                            transferResult(f, REL_OVERTOOK, null, targetOrder.get());
+                            targetOrder.set(isMaxOrder ? order : order + 1);
+                        } else {
+                            transferResult(f, REL_WAIT, null, targetOrder.get());
+                        }
+
+                    } else {
+                        final String msg = String.format("Skipped, FlowFile order was %d but current target is %d", order, targetOrder.get());
+                        logger.warn(msg + ". {}", new Object[]{f});
+                        transferResult(f, REL_SKIPPED, msg, targetOrder.get());
+                    }
+
+                });
+
+                if (previousTargetOrder != targetOrder.get()) {
+                    groupStates.put(stateKeyOrder, String.valueOf(targetOrder.get()));
+                    groupStates.put(STATE_UPDATED_AT.apply(groupId), String.valueOf(now));
+                }
+            });
+        }
+
+        private void transferResult(final FlowFile flowFile, final Relationship result, final String detail, final Integer expectedOrder) {
+            final Map<String, String> attributes = new HashMap<>();
+            attributes.put(ATTR_RESULT, result.getName());
+            if (expectedOrder != null) {
```

Comment:
Seems like when the flow file is to be transferred to SUCCESS then the attribute will not be set, but it may have already been set for the flow file if it had been transferred to WAIT. We should set the attribute to the actual order, or even better delete it to save memory since it won't be a useful attribute value anyway.